core/solver/batch_dispatch.hpp

```diff
@@ -0,0 +1,293 @@
+        : caller_{kernel_caller},
+          opts_{opts},
+          a_{matrix},
+          precon_{preconditioner},
+          logger_type_{logger_type}
+    {}
+
+    template <typename PrecType, typename BatchMatrixType, typename LogType>
+    void dispatch_on_stop(
+        const LogType& logger, const BatchMatrixType& amat, PrecType prec,
+        const multi_vector::uniform_batch<const device_value_type>& b_b,
+        const multi_vector::uniform_batch<device_value_type>& x_b)
+    {
+        if (opts_.tol_type == stop::ToleranceType::absolute) {
+            caller_.template call_kernel<
+                BatchMatrixType, PrecType,
+                device::stop::SimpleAbsResidual<device_value_type>, LogType>(
+                logger, amat, prec, b_b, x_b);
+        } else if (opts_.tol_type == stop::ToleranceType::relative) {
+            caller_.template call_kernel<
+                BatchMatrixType, PrecType,
+                device::stop::SimpleRelResidual<device_value_type>, LogType>(
+                logger, amat, prec, b_b, x_b);
+        } else {
+            GKO_NOT_IMPLEMENTED;
+        }
+    }
+
+    template <typename BatchMatrixType, typename LogType>
+    void dispatch_on_preconditioner(
+        const LogType& logger, const BatchMatrixType& amat,
+        const multi_vector::uniform_batch<const device_value_type>& b_b,
+        const multi_vector::uniform_batch<device_value_type>& x_b)
+    {
+        if (!precon_ ||
+            dynamic_cast<const matrix::Identity<value_type>*>(precon_)) {
+            dispatch_on_stop<device::Identity<device_value_type>>(
+                logger, amat, device::Identity<device_value_type>(), b_b, x_b);
+        } else {
+            GKO_NOT_IMPLEMENTED;
+        }
+    }
+
+    template <typename BatchMatrixType>
+    void dispatch_on_logger(
+        const BatchMatrixType& amat,
+        const multi_vector::uniform_batch<const device_value_type>& b_b,
+        const multi_vector::uniform_batch<device_value_type>& x_b,
+        log::BatchLogData<res_norm_type>& logdata)
+    {
+        if (logger_type_ == log::BatchLogType::simple_convergence_completion) {
+            device::batch_log::SimpleFinalLogger<res_norm_type> logger(
+                logdata.res_norms->get_values(),
+                logdata.iter_counts.get_data());
+            dispatch_on_preconditioner(logger, amat, b_b, x_b);
+        } else {
+            GKO_NOT_IMPLEMENTED;
+        }
+    }
+
+    /**
+     * Solves a linear system from the given data and kernel caller.
+     *
+     * Note: The correct backend-specific get_batch_struct function needs to be
+     * available in the current scope.
+     */
+    void apply(const MultiVector<ValueType>* const b,
+               MultiVector<ValueType>* const x,
+               log::BatchLogData<res_norm_type>& logdata)
+    {
+        const auto x_b = device::get_batch_struct(x);
+        const auto b_b = device::get_batch_struct(b);
+
+        if (auto amat =
```

Comment:
nit: I would put this into a `dispatch_on_matrix`, so that there is a clear chain of what factors into the dispatching.